Implements #19826 and #4571: Adds DbConnection.ProviderFactory and DbProviderFactories

FransBouma opened the pull request:
This issue (re)implements #19826: DbConnection.ProviderFactory, by simply returning the value returned by the virtual property DbConnection.DbProviderFactory. This way it's easier to bring back DbProviderFactories from .net full. 

The code was removed in commit 8150669, associated with issue #17905. 

cc @karelz @saurabh500 

(edit). This PR tracks now also the implementation of #4571. 

The public API is as follows: 
```cs
public static partial class DbProviderFactories
{
	public static DbProviderFactory GetFactory(string providerInvariantName)
	public static DbProviderFactory GetFactory(DataRow providerRow)
	public static DbProviderFactory GetFactory(DbConnection connection)
	public static void SetFactory<TFactory>(string name="", string providerInvariantName="", string description = "") where TFactory : DbProviderFactory
	public static void SetFactory(DbConnection connection, string name="", string providerInvariantName = "", string description = "")
	public static DataTable GetFactoryClasses()
}
```

The public API of the NetFX DbProviderFactories class is the same as the public API of this class.

Differences: 
- It doesn't auto-initialize itself, by default no factory/providers are registered.
- It can be manipulated using a public API at runtime. 

Usage to register a factory:
```cs
DbProviderFactories.SetFactory<SqlClientFactory>();
// or
DbProviderFactories.SetFactory(new SqlConnection());
```

The method accepts more parameters, so the developer can specify their own InvariantName, name and description if they want to. If they're not supplied, like the example above, they're obtained from the Type.

FransBouma commented:
Apparently a check failed but where it failed isn't related to the code I changed...

stephentoub commented on src/System.Data.Common/src/System/Data/Common/DbConnection.cs:
```diff
@@ -35,6 +35,8 @@ protected DbConnection() : base()
         /// </summary>
         protected virtual DbProviderFactory DbProviderFactory => null;
 
+        internal DbProviderFactory ProviderFactory=>DbProviderFactory;
```
Nit: spaces around the `=>`

stephentoub replied:
If it's important enough to add back an internal member because folks are using reflection to get at it, whether or not we "officially" support that we should have a test in place that similarly uses reflection to get it and verifies the result... otherwise it's likely in the future someone will come along and remove it. (For that reason a comment next to it would be good.)

stephentoub replied:
Also, @ericstj, is illink going to remove this?  @FransBouma would need to do something to prevent that, right?

FransBouma replied:
Will fix the space issue. 

The internal member is important as DbProviderFactories (the class I'll try to bring back next) is using this member on .NET full to obtain the factory in its method GetFactory(DbConnection), so to provide the same interface on .net core, this method is needed. It's currently indeed 'dead' as nothing references it. A comment is indeed needed here.

FransBouma replied:
I'll add a test for this tomorrow as well, next to the test in the other issue and will add a comment why it's there.

stephentoub commented:
@dotnet-bot test Innerloop OSX10.12 Debug x64 Build and Test please (https://github.com/dotnet/corefx/issues/19435)

FransBouma commented:
@karelz @saurabh500 
The test project of System.Data.Common seems to be misconfigured. It references the System.Data.Common.dll from:
"C:\Myprojects\corefx\bin\testhost\netcoreapp-Windows_NT-Debug-x64\shared\Microsoft.NETCore.App\9.9.9\System.Data.Common.dll"

While the System.Data.Common project builds its output to: ..\..\..\bin\AnyOS.AnyCPU.Debug\System.Data.Common\netcoreapp\

so the tests use the default one, not the one build in the same solution, ergo, changing something in the System.Data.Common project and then running the tests won't hit the new code, as far as I can tell, but the one in a different folder. 

Looking into the SqlClient test project/solution, I see System.Data.SqlClient is build to the WindowsNT folder, not the AnyOS one. Found this out as the test I wrote failed because the property I added wasn't there, obviously as the test references the original dll. 

However I have no idea how to fix this, or whether it's something in my configuration. I therefore can't test whether the test is OK, as it references a different assembly at runtime than the one in the same solution. 

Rather confusing, to be honest...

FransBouma commented:
I pushed the test to the PR, so I can check whether it runs in the CI... if it fails there too, it's not my system's corefx setup that's broken.

FransBouma commented:
Hmm test is failing in CI too but on a different line. Bizarre! Locally it fails on line 153 (the PropertyInfo isn't retrieved, property not found), in the CI it fails on line 155, the factory returned isn't an instance (which is really odd, considering the instance is a static property initialized with ctor call). 

I have no idea what's going on. Giving up for now...

FransBouma commented:
Ah, looking at the build of system.data.common on the command line, I see ILLinkTrimAssembly runs after build, which likely removes the property I am testing and therefore the test fails locally. Why it fails on a different line in the CI is weird (likely ILLinkTrimAssembly isn't running there?). Anyway, the removal of this property automatically has to be stopped first. No idea what to do about that, is there an attribute to add?

FransBouma commented:
Found this page: https://github.com/dotnet/corefx/blob/master/Documentation/building/windows-instructions.md which describes what I found is already known. Good to know, would have been nice if it was known up front. I've updated https://github.com/dotnet/corefx/wiki/Build-and-run-tests to link to that page directly for people who think about running tests within visual studio so they'll be guided to that page as it's not directly obvious to go there.

saurabh500 commented:
The two legs which are passing and are actually running the tests are 
`Innerloop netfx Windows_NT Debug x86 Build and Test` and `Innerloop netfx Windows_NT Release x64 Build and Test` 
These two legs are running the tests against .Net Framework where the internal property is available. The CoreFx test are failing because of ILLink. 

I don't know if ILLink cleanup can be prevented by adding an attribute. @Karelz in case you can get someone to help with this information. 

@FransBouma In case ILLink cleanup cannot be overridden, this change will need to come in with the bigger work around `DbProviderFactories` that you intend to do where the internal is used and will not be cleaned up

FransBouma commented:
@saurabh500 @karelz I agree, this line of code can be added better with code which actually uses it, so there's no problem with ILLink. I've planned to port over DbProviderFactories today, so I'll push that to this PR so this line of code will then be needed and won't be weeded out by ILLink.

FransBouma commented:
Ok, adding a test for DbProviderFactories to the System.Data.Common.Tests project fails, as, like I said before, it doesn't see the assembly built by the System.Data.Common project, so it doesn't see my added DbProviderFactories class... 

I have no idea what to do now. Sorry, but this is very frustrating. @karelz @saurabh500

FransBouma commented:
I think the problem is that the build configuration for System.Data.Common.Tests is for netstandard-debug, and DbProviderFactories isn't in .net standard. But I'm guessing here, no clue how all these test projects are setup and I rather not alter things there or add a new test project just for this class...

FransBouma commented:
Tried everything, but no dice: the Tests project refuses to see the added public DbProviderFactories class. I need to run a basic test now before I can proceed further, but that's not possible now. I have no idea how you people do this, I don't care if I have to use notepad and a .cmd to build, but with even those bare bones tools I don't know where to start with this. The code itself is no problem, it's only the tooling around it that are in terrible shape, sorry. 

I had to add exception strings so I tried to add them to the Strings.resx file, but inside vs2017 this crashed the IDE, with reproducible steps, likely due to a mismatch between System.Runtime (as it popped up an exception related to that before crashing). I reported this issue. Adding the string in the resx file by hand in notepad++ doesn't get picked up even if I reload the project, as it's not added to `SR`. I see SR is generated at build, but it doesn't pick up my additions. Viewing the Strings.resx file in vs does work, it shows my added strings, but SR doesn't get properties for it. So after fighting with crashing IDEs for a while and way to add to SR I added the exception strings in the DbProviderFactories class. This has to be ported over to the resources later, but first things have to **work**, and that's already a serious challenge.

saurabh500 commented:
@fransbouma didbyou modify the files in system.data.common/ref to expose public api so that test can access the API?

FransBouma commented:
@saurabh500 No haven't altered anything there. Looking at it indeed, I should add a class there with the public methods, thanks for this tip :) I've added the class and api to system.data.common/ref/system.data.common.cs: 

```cs
    public static partial class DbProviderFactories
    {
        public static DbProviderFactory GetFactory(string providerInvariantName) { throw null; }
        public static DbProviderFactory GetFactory(DataRow providerRow) { throw null; }
        public static DbProviderFactory GetFactory(DbConnection connection) { throw null; }
        public static void SetFactory<TFactory>(string name="", string providerInvariantName="", string description = "") where TFactory : DbProviderFactory { throw null; }
        public static void SetFactory(DbConnection connection, string name="", string providerInvariantName = "", string description = "") { throw null; }
        public static DataTable GetFactoryClasses() { throw null; }
    }
```
that compiles fine, but the tests project still refuses to see the class, building on command line, rebuilding everything using build-managed etc. doesn't help... DbProviderFactories still isn't a known type and I can't use it... 

I have this silly code: 
```cs
// Licensed to the .NET Foundation under one or more agreements.
// See the LICENSE file in the project root for more information.

using System.Reflection;
using System.Data.Common;
using Xunit;

namespace System.Data.Common
{
    public class DbProviderFactoriesTests
    {
        [Fact]
        [Trait("Issue", "I19826")]
        public void InitializationTest()
        {
            DataTable initializedTable = DbProviderFactories.GetFactoryClasses();
        }
    }
}
```
This doesn't compile, "`DbProviderFactories` isn't known in the current context.". If I type: `System.Data.Common.DbProviderFactory` and press F12 on DbProviderFactory, I go to the netstandard assembly, for .net full. 4.6.1. Not sure if that's resharper's fault, anyway, the compiler doesn't build this code, and I can't run any tests against the class. Not sure if I have to alter another file somewhere...

FransBouma commented:
Forgot to mention: if I look into the assembly in corefx\bin\ref\System.Data.Common\4.2.0.0\netcoreapp, the class is there, so something else is going on in vs2017/project. Not sure if I have to install 15.3 preview for this, as it fails on the command line too.

FransBouma commented:
I've now implemented everything for DbProviderFactories, commented out the tests for that for now and the test for DbConnection.ProviderFactory which failed earlier in this PR now succeeds, so that's at least solved :) All that's needed now is somehow to get the compiler to see DbProviderFactories and we're almost done :)

FransBouma commented:
I won't push the code as of yet as the code will then end up in this PR and potentially merged, so I'll wait till the test stuff is resorted.

saurabh500 commented:
Can you push the code to another branch of your fork so that I can look at the problems with the test?

FransBouma commented:
@saurabh500 Thanks, I'll try that. :)

FransBouma commented:
@saurabh500  I've  pushed it to this branch: https://github.com/FransBouma/corefx/tree/Implementation_DbProviderFactories

Implementation is there (exception strings should be moved to resources) and tests should be written. Hopefully you can find out what's wrong! Thanks in advance :)

karelz commented:
I see long discussion with lots of details, and I sense some frustration. @FransBouma is there something that doesn't work well? Can you please summarize it for me, so that I can ask further? Thanks!

saurabh500 commented:
@karelz The problem statement here is essentially
"If a new API needs to be added to System.Data.Common and tested from System.Data.Common.Tests, then how does one go about doing this? The tests are referencing the netstandard.dll to resolve the Types and if a new type is added to System.Data.Common, what's the right way to test this new Type addition. The test is not able to resolve the new 
A new API implementation was added in the form of a new Type and related methods and the reference code was updated with the new type and methods."

Apart from this, I have tried adding a direct reference to System.Data.Common.csproj in the tests and that leads to other build issues.

ericstj commented:
If you're adding an API to the netcoreapp version of System.Data.Common then you should make sure you have a netcoreapp configuration of System.Data.Common and add your tests there.

With System.Data.Common being part of netstandard, you cannot add API to the nestandard configuration of System.Data.Common.  The netstandard tests for System.Data.Common can only test the netstandard API.

karelz commented:
It will be eventually part of our docs: https://github.com/dotnet/corefx/wiki/Contributions#expose-api-in-corefx
This is a bit advanced scenario, so the docs are suboptimal.

FransBouma commented:
@ericstj changing that line to netcoreapp didn't solve it unfortunately, I kept getting the same build issues. I agree, the test project that was present tested against netstandard, so a new test project or a configuration change to make the change to test against netcoreapp should be made, however I couldn't come up with one, mostly because I'm still inexperienced in this area. 

I hope it gets sorted so I can conclude this work :)

FransBouma commented:
@karelz it's mostly about the assumption the test project was able to test against netcoreapp but it was testing against netstandard and I couldn't find a way to make that change. I still have no idea how to configure it properly, so the work is effectively halted till this is solved and that's a little frustrating, as most time spent on this was on fighting the tooling/configurations, not thinking about what code to write.

FransBouma commented:
> To create a netcoreapp configuration of the tests, update this: https://github.com/FransBouma/corefx/blob/7102d7c97fef3d0754bf8dff5472e4420ed00ce1/src/System.Data.Common/tests/Configurations.props#L5
> 
> Then make the appropriate additions to the test in a netcoreapp specific source file conditioned appropriately, or if-def'ed in a common source file.

It would be *great* if there were instructions what to do. The above describes there's something wrong and there have to be taken actions, but which ones is a *complete blur* to me. I can't be the only one who will run into this problem and it's going to be a problem for users of CoreFX too who in their own code have tests for netstandard and tests for corefx parts of their code. 

So back to trial/error driven development I guess till someone fixes this... :/ 

My wild guess would be:
- add netcoreapp to the line specified, keep the other line ? 
- add an #ifdef netcoreapp around my test ? 

cc @karelz

FransBouma commented:
(Sorry for the flood!) I just thought about this: Yesterday I tried to find a project that tested against netcoreapp and netstandard within corefx's list of projects but couldn't find any (I picked a couple at random, so didn't try all as there are many), I did this to see how these are set up so I could copy that setup to get things fixed. Is there such a project in the repo so I could look at that and apply the differences to this project so things get unblocked?

karelz commented:
@FransBouma very sorry, but you have stumbled upon expert-area. These things are much better with new infra as of January, but we still lack clear description and guidance / examples what exactly to do. It would be best to wait for experts to provide guidance to avoid more wasted time on your side.

@weshaggard @mellinoe @ericstj can you please advice? (ideally verified locally)
Or even better, can you please sketch rough how-to into the [docs](https://github.com/dotnet/corefx/wiki/Contributions#expose-api-in-corefx)? ... I think we used to have some docs around these things, but I don't know if they are still valid + they were very long and devs had to become experts first. I think we need simple how-tos for most common cases as listed in the [docs](https://github.com/dotnet/corefx/wiki/Contributions#expose-api-in-corefx)
This has been a bit magic area even for developers on our team, so it would be nice to clarify it (after we're done with 2.0 high-pris). Thanks!

FransBouma commented:
I've figured it out with the help of this: https://github.com/dotnet/corefx/pull/19147#issuecomment-301352244 what to do.
(I've removed the DbProviderFactories type from the `rev` part, as it's not part of netstandard)
To make the tests compile for netcoreapp and not for netstandard, the following steps have to be taken: 

* add `netcoreapp;` to the configurations.props file in the Tests folder, below `netstandard;`
* add to the system.data.common.tests.csproj, below the 2 netstandard configurations: 
```xml
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'netcoreapp-Debug|AnyCPU'" />
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'netcoreapp-Release|AnyCPU'" />
```
* add to the system.data.common.tests.csproj, another item group:
```xml
  <ItemGroup Condition="'$(TargetGroup)' == 'netcoreapp'">
    <Compile Include="System\Data\Common\DbProviderFactoriesTests.cs" />
  </ItemGroup>
```

That's basically it. Doing `build-tests` will now build two outputs, one with and one without the DbProviderFactories tests file. I'll rename the file properly with 'netcoreapp' in the name like in the link referenced above, so it's clear it's for .netcoreapp. 

Haven't tested this in vs2017 though. Not sure if the 15.2 tooling picks up conditionals. No more time today, will try tomorrow. :)

cc: @karelz @saurabh500

karelz commented:
Cool, happy to see progress. It seems right - @mellinoe @weshaggard @ericstj do you have additional suggestions?

ericstj commented:
Nope, that's exactly what I was suggesting above: https://github.com/dotnet/corefx/pull/19885#issuecomment-302827302.  Sorry it wasn't clear and you burned some cycles on it.  I'll review the update once you push it.

The better way to change these configurations is to update `Configurations.props`, then rerun `msbuild /t:UpdateVSConfigurations` from the root.  That will make sure project files are in-sync with Configuration.props as well as regenerate all sln files (which is likely what's causing your VS build to be a little off).  You may run into a failure running UpdateVSConfigurations which I noticed was recently broken: https://github.com/dotnet/buildtools/pull/1509

You can always test from the commandline by running msbuild /t:BuildAndTest on the test project.

Just for future information, there are quite a few tests that do this.  Here are some examples:
https://github.com/dotnet/corefx/blob/6dd451f51451a7d0ceea6104b51bd17005e9a0e6/src/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Tests.csproj
https://github.com/dotnet/corefx/blob/b58a7977d2776bf2d02e29c3ba4b7597720ee7cb/src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj

ericstj commented:
All of this is covered in the docs here: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/project-guidelines.md#aggregate-build-properties

We'd be happy to take any changes to this doc or cross-reference from other docs to make this clearer.

FransBouma commented:
@ericstj Thanks for the info :)

I tried rebuilding the solution files with the command suggested, but get a glorified error: 
```
C:\Myprojects\corefx>msbuild /t:UpdateVSConfigurations
Microsoft (R) Build Engine version 15.1.1012.6693
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 23-5-2017 11:32:23.
Project "C:\Myprojects\corefx\build.proj" on node 1 (UpdateVSConfigurations target(s)).
UpdateVSConfigurations:
  Updating configurations for projects ...
C:\Myprojects\corefx\build.proj(66,5): error MSB4018: The "UpdateVSConfigurations" task failed unexpectedly.\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018: System.InvalidOperationException: Sequence contains no elements\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.UpdateVSConfigurations.EnsureP
rojectGuid(ProjectRootElement project)\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.UpdateVSConfigurations.Execute
()\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.B
ackEnd.ITaskExecutionHost.Execute()\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTas
k>d__26.MoveNext()
Done Building Project "C:\Myprojects\corefx\build.proj" (UpdateVSConfigurations target(s)) -- FAILED.


Build FAILED.

"C:\Myprojects\corefx\build.proj" (UpdateVSConfigurations target) (1) ->
(UpdateVSConfigurations target) ->
  C:\Myprojects\corefx\build.proj(66,5): error MSB4018: The "UpdateVSConfigurations" task failed unexpectedly.\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018: System.InvalidOperationException: Sequence contains no elements\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.UpdateVSConfigurations.EnsureP
rojectGuid(ProjectRootElement project)\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.UpdateVSConfigurations.Execute
()\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.B
ackEnd.ITaskExecutionHost.Execute()\r
C:\Myprojects\corefx\build.proj(66,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTas
k>d__26.MoveNext()

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.54
```
Looks like some iterator fails to obtain a list to process. This fails almost immediately so it's not due to the changes I made to the system.data.common file. This fails with the same error when I switch to the master branch and update from master/upstream. 

I have the feeling everything is a bit broken and adding code now is the worst possible moment... :/

FransBouma commented:
Ok, to recap, as things are a bit problematic at the moment:

- Code for the DbProviderFactories class is written and unless I made glorified mistakes, should work
- Strings in the exceptions should be moved to the Strings.resx file, this is currently blocked by a problem within VS2017 which crashes when editing the resource file. 
- An API review has to be performed for the additional method 'SetFactory' and its overloads.
- A facade has to be created for SetFactory for netfx so the method being in NetStandard2.1 works on netfx too
- Tests have to be written to make sure it works on all platforms. This is currently halted by infra/configuration issues. 
- SqlConnection.DbProviderFactory has to be implemented (#19825, PR submitted (#19884), ready to be merged)

As I use visual studio, it has been a real struggle to get things up and running. It's not the code, that's straightforward, but getting to know what to do with the infrastructure and then failing to make it work due to blind spots in the system has taken a lot of time ( so much so in fact that I'll halt development on this for now till tools work again). 

I have learned CoreFx team doesn't use visual studio, which is understandable considering the multi-platform nature, however for a person like me who uses an IDE primarily for C#/.NET development it's a bit of a struggle to make things work within a command line oriented system. I can switch to command line development using e.g. vscode, it's just a matter of how to e.g. debug code if things go wrong and how to add an API that's not present in the current config, like I had to do here: add a test project for netcoreapp while the only test project was for netstandard). 

As I can't get it to work within vs2017 at the moment, I can't proceed. This is a bit of a shame as I've spent a lot of time getting here, trying to work around the issues I ran into, but c'est la vie. 

I'll wait for the infra/tooling to become usable for me to add code within vs so I can add the tests and submit an api review. Thanks.

cc @karelz @saurabh500  @ericstj

FransBouma commented:
Doing a simple `build-tests` also fails to build the tests, even with the adjustments suggested, so there's something else that needs adjustment that's currently unclear. More guidance for how to set this up is dearly needed. (additionally to: how to run a single test, how to debug a test from the command line)

karelz commented:
@ericstj @mellinoe @weshaggard do you have bandwidth to take @FransBouma's change and try it locally the same way he does? (if you're busy with Ask mode bugs, this can wait for next week, when we are in Escrow).

FransBouma commented:
@ericstj 
> You may run into a failure running UpdateVSConfigurations which I noticed was recently broken: dotnet/buildtools#1509

I completely overlooked this line, sorry. You warned me up ahead and I reported exactly what you warned for. Sorry for that! Will try the latest repo source tomorrow.

divega commented:
@FransBouma, @saurabh500, @karelz and others interested, since all the code isn't here yet, I went ahead and made a few comments regarding the API design at https://github.com/FransBouma/corefx/commit/2edb9855d1c849ea6855ee6b1ded565375fe4ecc.

FransBouma commented:
Compiling the sln with conditionals on the ItemGroup doesn't work inside vs2017, it e.g. also doesn't compile System.Text.RegularExpressions here. This is currently broken, so no time to spend on that anymore, till that part of the tooling is updated. 

So the tests have to be compiled on the command line, run on the command line and debugged on the command line (with a while(!Debugger.Attached()) {Sleep(100);} wait loop to attach a debugger) for now.

FransBouma commented:
When trying msbuild /t:BuildAndTest in root of System.Data.Common, it shows something odd (after making all the config changes). 

```
Project "C:\Myprojects\corefx\src\System.Data.Common\src\System.Data.Common.csproj.metaproj" (3) is building "C:\Myprojects\corefx\src\System.Data.Common\src\System.Data.Common.csproj" (32) on node 1 (BuildAndTest target(s)).
CoreResGen:
  No resources are out of date with respect to their source files. Skipping resource generation.
GenerateResourcesSource:
Skipping target "GenerateResourcesSource" because all output files are up-to-date with respect to the input files.
GenerateAssemblyInfo:
Skipping target "GenerateAssemblyInfo" because all output files are up-to-date with respect to the input files.
CoreCompile:
Skipping target "CoreCompile" because all output files are up-to-date with respect to the input files.
OpenSourceSign:
Skipping target "OpenSourceSign" because all output files are up-to-date with respect to the input files.
CopyFilesToOutputDirectory:
  System.Data.Common -> C:\Myprojects\corefx\bin\AnyOS.AnyCPU.Debug\System.Data.Common\netcoreapp\System.Data.Common.dll
Done Building Project "C:\Myprojects\corefx\src\System.Data.Common\src\System.Data.Common.csproj" (BuildAndTest target(s)).
```

However, the test project build fails, as it refers to: 
```
Project "C:\Myprojects\corefx\src\System.Data.Common\tests\System.Data.Common.Tests.csproj.metaproj" (2) is building "C:\Myprojects\corefx\src\System.Data.Common\tests\System.Data.Common.Tests.csproj" (33) on node 1 (BuildAndTest target(s)).
CoreCompile:
  C:\Windows\Microsoft.NET\Framework\v4.0.30319\csc.exe /noconfig /nowarn:0168,0169,0414,0219,0649,1701,1702 /nostdlib+
   /warn:4 /define:DEBUG;TRACE;netcoreapp;DEBUGRESOURCES;SIGNED /reference:C:\Myprojects\corefx\bin/ref/netcoreapp/Microsoft.CSharp.dll 
   /reference:C:\Myprojects\corefx\bin/ref/netcoreapp/Microsoft.VisualBasic.dll 
   /reference:C:\Myprojects\corefx\bin/ref/netcoreapp/Microsoft.Win32.Primitives.dll 
   ...
   /reference:C:\Myprojects\corefx\bin/ref/netcoreapp/System.Data.Common.dll
```   
which obviously isn't the same dll. Here my knowledge ends, as I don't know why this is. But I think it's the crux of the matter why the test build fails here: the netcoreapp build refers to a dll that's not the same as is build by the sln earlier (as that one is in AnyCPU).

FransBouma commented:
(CC @karelz @saurabh500 ) for previous message

karelz commented:
@weshaggard @mellinoe @ericstj can one of you please try @FransBouma's changes locally with his steps? He invested already a lot of time trying to fight our infra (1 week of full-time work), but without much success. We should try to help him get unblocked more proactively. Thanks!

Key goals: 1. Get @FransBouma unblocked. 2. Understand what breaks on his path of adding .NET Core specific API, fix it & document it in new contributor docs to prevent such situation to come up again.

If you don't have time, please let me know and I will try to ask someone else. cc @danmosemsft

weshaggard commented:
>When trying msbuild /t:BuildAndTest in root of System.Data.Common, it shows something odd (after making all the config changes).

Unfortunately you cannot run the BuildAndTest target from the root which tries to run it on the solution file. The BuildAndTest target only works on the test project itself. Can you try to do run it on the test project itself.

Note: As a hint if you see the ".metaproj" extension that is a generated wrapper project that comes when trying to build a solution file from the command line.

weshaggard commented:
Another think I noticed is that you didn't update the ref for System.Data.Common to include the new types which means that the tests will not see the new APIs.

FransBouma commented:
@weshaggard 
> Unfortunately you cannot run the BuildAndTest target from the root which tries to run it on the solution file. The BuildAndTest target only works on the test project itself. Can you try to do run it on the test project itself.

Ah, will try tomorrow, thanks for the heads up!

> Another think I noticed is that you didn't update the ref for System.Data.Common to include the new types which means that the tests will not see the new APIs.

No I didn't do that indeed, as I was under the assumption that that ref was a proxy for the netstandard ref (as the test project was previously only testing against netstandard).

FransBouma commented:
@weshaggard That was it indeed! I added the DbProviderFactories type in a separate cs file to ref (as System.Data.Common.netcoreapp.cs, same as e.g. System.Text.RegularExpressions does). No conditional compile in the csproj tho (as text.regularexpressions doesn't do that too). Isn't it necessary to have a conditional on the ItemGroup for netcoreapp for that file in the ref project? As it isn't needed for the build for other platforms (it won't hurt of course, but I can imagine you might not want it to be there in the ref for platforms which doesn't have the type yet or might never get it).

karelz commented:
> this info bit about 'ref' is likely a key issue to document on the contributors wiki (or other docs).

Agreed. Please add it there - where you think it fits best.
IMO we should link your PR once it is merged as example of all things which are necessary.

weshaggard commented:
>Isn't it necessary to have a conditional on the ItemGroup for netcoreapp for that file in the ref project?

FransBouma commented:
> You should just put the new APIs direct in System.Data.Common.cs file under ref there, especially given it isn't conditioned on anything there is no good reason to split it out.

Actually it is currently, it's part of .netcore, not .netstandard. So isn't it in the same boat as the .netcore specific additions to e.g. system.text.regularexpressions? I had to add the class as a conditional to the src project. 

My point is that if test code refers to the type and the build is for NetFX, it should fail at build time, but from what I understand it doesn't as the ref contains the type so it always builds and fails at runtime as the type can't be forwarded (in this highly particular case it can, but that's a coincidence. For the sake of my point, let's assume NetFX doesn't contain DbProviderFactories. 

Does that make sense? :)

weshaggard commented:
>Actually it is currently, it's part of .netcore, not .netstandard. So isn't it in the same boat as the .netcore specific additions to e.g. system.text.regularexpressions? I had to add the class as a conditional to the src project.

The RegularExpressions example is split unnecessarily as we don't build a netfx ref, it is probably that way based on history of the file but it could be merged. However for System.Data.Common you are correct that we do build a netfx reference so any new APIs need to be correctly conditioned, however it is my understanding that this isn't a new API but instead it is exposing a API that already exists in netfx so you can keep things merged in your case as well.

divega commented:
@weshaggard the type exists in .NET Framework and Mono but we are adding new API to it. See https://github.com/dotnet/standard/issues/356 for explanation (the specifics of the API may currently not match between this PR and that draft proposal).

weshaggard commented:
>@weshaggard the type exists in .NET Framework and Mono but we are adding new API to it. See dotnet/standard#356 for explanation (the specifics of the API may currently not match between this PR and that draft proposal).

In that case I would put all the stuff that is common into System.Data.Common.cs and only use the netcoreapp partial class file for things that are new and unique to .NET Core.

FransBouma commented:
> In that case I would put all the stuff that is common into System.Data.Common.cs and only use the netcoreapp partial class file for things that are new and unique to .NET Core.

So the ref project has a .netcoreapp.cs file with only the type+method(s) which are specific for netcoreapp, and all other methods/type are added to the system.data.common.cs file? 

I'll do that. I'll also include the api suggestions from @divega  regarding SetFactory -> ConfigureFactory. The api proposal in dotnet/standard#356 doesn't contain overloads of COnfigureFactory. I'll post in that thread when I have that implemented. Hopefully some commits later today :)

FransBouma commented:
There's one task left, if everything passes, which is moving the hard-coded strings for exceptions to the string resource file. I'll do that after the CI completes without errors :)

FransBouma commented:
Ok, mac tests failed on a test in security. It took a tremendously long time, so maybe that's it. I'll now move the strings (maybe this evening otherwise tomorrow) and will commit again, so the CI will run everything again anyway. 

I also noticed I missed a test, so I'll add that too. ALmost done!

divega commented on src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs:
```diff
@@ -0,0 +1,187 @@
+        private static void RegisterFactoryInTable(Type factoryType, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(factoryType, nameof(factoryType));
+            string invariantNameToUse = string.IsNullOrWhiteSpace(providerInvariantName) ? factoryType.Namespace : providerInvariantName;
```
It seems in the two places where we call this private method we don't allow `providerInfariantName` to be null. Do we still need this fallback to the namespace?

divega commented on src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs:
```diff
@@ -0,0 +1,187 @@
+        
+        private static void RegisterFactoryInTable(Type factoryType, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(factoryType, nameof(factoryType));
```
Probably not worth checking the argument again in private methods.

FransBouma replied:
It's a leftover from the default parameter values, good catch! 

What should I do: 
1) add an overload which doesn't requires providerinvariantname (so fall back to namespace there)
2) remove the check 
3) leave it. 

I don't know what the general policy is in this case inside MS: always check your input no matter the call path, or don't check if the call path already has checked things...

divega replied:
@FransBouma I am a bit lost. It seems the fallback to namespace is currently never used. Doesn't that mean it can just be removed?

It this was one of the repos over at https://github.com/aspnet I would feel more confident suggesting the private method doesn't need to have the check if you are sure it is checked in all the public entry points, but there is some variability, e.g. you could choose to have a `Debug.Assert()` instead. So I would start with @karelz for corefx guidance and @saurabh500 for current practices.

And while we are on it, can we switch to `var` everything that can be switched to `var` in this code?

FransBouma replied:
@divega At the moment it appears it's not used indeed. It's fine by me to remove it, or replace it with a Debug.Assert, whatever makes you all happy :)

> And while we are on it, can we switch to var everything that can be switched to var in this code?

I'd love to, I followed the coding guidelines which explicitly stated 'var' was out of the question and should be avoided, so I used the types instead :)

saurabh500 replied:
Looking at the ADO.Net code, it uses Debug.Assert over multiple checks in the same code path. It would be good to avoid many checks in the same code path for the same condition. 
No `var` please.

karelz replied:
FYI: var coding style: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md ... I personally suggest to use var only if it makes the line clearly more readable (i.e. not for string), and if it adheres to rule [10] in the link (i.e. the type is inferable from the line - think FindInFiles usages)

Also let's not mingle (larger) formatting changes with real changes (see [notes](https://github.com/dotnet/corefx/wiki/Contributions#small-change)).

FransBouma replied:
@saurabh500 so this particular check should be a Debug.Assert?

FransBouma replied:
I'll refactor the API a bit. I'll add an overload which only accepts a type, and will fall back to the namespace. This change is discussed further in the review of this API here: https://github.com/dotnet/standard/issues/356

karelz replied:
We should not track anything in dotnet/standard repo. All API reviews should happen in CoreFX. Standard is about making existing things as part of standard. It should be the last step of everything. I am still confused if we are porting existing APIs from Desktop, or if we are adding brand new overloads/APIs. @FransBouma can you please clarify?

saurabh500 commented on src/System.Data.Common/System.Data.Common.sln:
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 14
-VisualStudioVersion = 14.0.25420.1
+# Visual Studio 15
```
Can we leave these versions unchanged?

FransBouma replied:
Sure, I just edited in vs2017. It loads OK in 2015 tho, but I'll revert the change.

saurabh500 commented on src/System.Data.Common/System.Data.Common.sln:
```diff
@@ -26,8 +26,8 @@ Global
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
-		{B473F77D-4168-4123-932A-E88020B768FA}.Debug|Any CPU.ActiveCfg = netstandard-Debug|Any CPU
-		{B473F77D-4168-4123-932A-E88020B768FA}.Debug|Any CPU.Build.0 = netstandard-Debug|Any CPU
+		{B473F77D-4168-4123-932A-E88020B768FA}.Debug|Any CPU.ActiveCfg = netcoreapp-Debug|Any CPU
```
Shouldn't netcoreapp-Debug|Any CPU be added in addition to `netstandard-Debug|Any CPU`?
Was this modified manually or did VS overwrite the existing files?

FransBouma replied:
This is done by vs2017. It changes this when you change the target build of a project in Build->Configuration manager. I had to do this, as the default was netstandard, so the code wouldn't  build (as the project system didn't honor the conditional include of the netcoreapp cs file, as the tooling for that is still being worked on (I use 15.2)

saurabh500 commented on src/System.Data.Common/src/System/Data/Common/DbConnection.cs:
```diff
@@ -35,6 +35,8 @@ protected DbConnection() : base()
         /// </summary>
         protected virtual DbProviderFactory DbProviderFactory => null;
 
+        internal DbProviderFactory ProviderFactory => DbProviderFactory; // This property is required for DbProviderFactories' API and will appear as 'dead' code till that class has been ported.
```
You could get rid of the comment now that the consumer of this property is being brought in

FransBouma replied:
You're right! will remove it.

FransBouma commented:
Should be all good now. For API work, I'd like to refer to the review issue: https://github.com/dotnet/standard/issues/356. I've added a description of the API, with per method it's purpose and what it does. All remarks made are now addressed.

karelz commented:
@FransBouma can you please help me understand if your PR is adding brand new API? If that's correct, we need to have issue in CoreFX tracking [formal API review](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) -- I provided more details in https://github.com/dotnet/standard/issues/356#issuecomment-307578096.

If this PR has multiple parts, beside API addition, I would suggest to split it off.

We should sync with @divega @saurabh500 to create best plan forward.

FransBouma commented:
Part of this PR is ported from netfx, but in netfx it autoinitializes (from machine.config) which isn't available on corefx, so I added 2 methods (and overloads) for initialization for that (the ConfigureFactory() methods). These 2 methods (and their overloads) are specific for corefx at the moment, and plans are to add those to netstandard.

I can't split it off, as the code ported from netfx can't initialize itself.

karelz commented:
I see, thanks for info. In that case: Let's create new API addition issue. Let's put just the API surface there for review. On Monday I will ask @divega and @saurabh500 to review it. If it is ok, we can bring it up in Tuesday API review meeting. Once approved, we are good on the API in CoreFX repo. (Standard will be much later, afetr we decide to port it to Desktop eventually - think months at best)

Does it make sense? Any further questions/suggestions/push backs?

FransBouma commented:
Sorry @karelz but I'm not going to write yet another document / post for yet another review pipeline. I don't even know what it has to contain. This issue: https://github.com/dotnet/standard/issues/356 and this comment https://github.com/dotnet/standard/issues/356#issuecomment-307552750 contains all the information needed. 

I've spent so much time on this to make everything as you want it to be, read guidelines, etc, for 200 lines of code which are very straightforward. the most time per line of code I've spend in a long time, and I'm done with it. 

This PR contains the code for DbProviderFactories and initialization logic for corefx. Including tests. If you want it, you can merge it. If you want to change it, change it. I'm out.

divega commented:
@FransBouma I went ahead and created a new issue specific to the new API that doesn't come from desktop as @karelz requested. I have copied the full text from your post summarizing the new API, so nothing is lost :smile:

divega commented:
@FransBouma BTW, the new overloads that only take the type make sense given the fallback logic. It seems ver usable. The only question I personally have about the API is whether we want to go instead with the minimal API that enables all scenarios, which is only one method. But that is for the API discussion.

FransBouma commented:
> I went ahead and created a new issue specific to the new API that doesn't come from desktop as @karelz requested. I have copied the full text from your post summarizing the new API, so nothing is lost 

@divega thanks for that :) 

> The only question I personally have about the API is whether we want to go instead with the minimal API that enables all scenarios, which is only one method. But that is for the API discussion.

Will continue there, thanks

karelz commented:
Marking the issue as blocked on API review.

karelz commented:
Still blocked :(, I pinged @divega on the API review issue #20903

FransBouma commented:
I closed this PR as the final code will use a fresh PR as it's an implementation of the final API after review. This makes it easier to review the PR's changes otherwise the history of this PR with 80+ comments makes it a bit difficult. I'll refer to this PR in the new PR for reference. Code is almost done (except a few minor tests), I expect it later this week. 

Ref: https://github.com/dotnet/corefx/issues/20903#issuecomment-342876016
